src/analyserqueue.cpp

```diff
@@ -159,61 +162,47 @@ TrackPointer AnalyserQueue::dequeueNextBlocking() {
 }
 
 // This is called from the AnalyserQueue thread
-bool AnalyserQueue::doAnalysis(TrackPointer tio, const Mixxx::SoundSourcePointer& pSoundSource) {
-    int totalSamples = pSoundSource->length();
-    //qDebug() << tio->getFilename() << " has " << totalSamples << " samples.";
-    int processedSamples = 0;
+bool AnalyserQueue::doAnalysis(TrackPointer tio, Mixxx::AudioSourcePointer pAudioSource) {
 
     QTime progressUpdateInhibitTimer;
     progressUpdateInhibitTimer.start(); // Inhibit Updates for 60 milliseconds
 
-    int read = 0;
+    Mixxx::AudioSource::size_type progressFrameCount = 0;
     bool dieflag = false;
     bool cancelled = false;
-    int progress; // progress in 0 ... 100
-
     do {
         ScopedTimer t("AnalyserQueue::doAnalysis block");
-        read = pSoundSource->read(kAnalysisBlockSize, m_pSamplesPCM);
+
+        DEBUG_ASSERT(progressFrameCount < pAudioSource->getFrameCount());
+        const Mixxx::AudioSource::size_type readFrameCount = pAudioSource->readSampleFramesStereo(kAnalysisFrameCount, &m_sampleBuffer[0], m_sampleBuffer.size());
+        progressFrameCount += readFrameCount;
+        DEBUG_ASSERT(progressFrameCount <= pAudioSource->getFrameCount());
 
         // To compare apples to apples, let's only look at blocks that are the
         // full block size.
-        if (read != kAnalysisBlockSize) {
-            t.cancel();
-        }
-
-        // Safety net in case something later barfs on 0 sample input
-        if (read == 0) {
-            t.cancel();
-            break;
-        }
-
-        // If we get more samples than length, ask the analysers to process
-        // up to the number we promised, then stop reading - AD
-        if (read + processedSamples > totalSamples) {
-            qDebug() << "While processing track of length " << totalSamples << " actually got "
-                     << read + processedSamples << " samples, truncating analysis at expected length";
-            read = totalSamples - processedSamples;
-            dieflag = true;
-        }
-
-        SampleUtil::convertS16ToFloat32(m_pSamples, m_pSamplesPCM, read);
-
-        QListIterator<Analyser*> it(m_aq);
-
-        while (it.hasNext()) {
-            Analyser* an =  it.next();
-            //qDebug() << typeid(*an).name() << ".process()";
-            an->process(m_pSamples, read);
-            //qDebug() << "Done " << typeid(*an).name() << ".process()";
+        if (kAnalysisFrameCount == readFrameCount) {
```

Comment:
Ah, I can answer myself: Yes.  I have mixed up progressFrameCount and readFrameCount.